File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.java
import java.util.*;

class Solution {
    public String shortestBeautifulSubstring(String s, int k) {
        List<Integer> ones = new ArrayList<>();

        // store positions of all 1s
        for (int i = 0; i < s.length(); i++)
            if (s.charAt(i) == '1')
                ones.add(i);

        // Not enough 1s
        if (ones.size() < k)
            return  "";

        String answer = "";

        // try every possible starting 1
        for (int i = 0; i + k - 1 < ones.size(); i++) {
            int start = ones.get(i);
            int end = ones.get(i + k - 1);

            String candidate = s.substring(start, end + 1);

            if (answer.isEmpty()
                    || candidate.length() < answer.length()
                    || (candidate.length() == answer.length()
                        && candidate.compareTo(answer) < 0)) {
                            answer = candidate;
            }
        }   

        return answer;     
    }
}

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        ones = []

        # store positions of all 1s
        for i in range(len(s)):
            if s[i] == '1':
                ones.append(i)

        # Not enough 1s
        if len(ones) < k:
            return ""

        answer = ""

        # try every possible starting 1
        for i in range(len(ones) - k + 1):
            start = ones[i]
            end = ones[i + k - 1]

            candidate = s[start:end + 1]

            if (answer == "" 
                    or len(candidate) < len(answer) 
                    or (len(candidate) == len(answer) and candidate < answer)):
                answer = candidate

        return answer
